### PR TITLE
En 4331/integrate system vm meta chain

### PR DESCRIPTION
**Pull request by sasurobert**
The document with the specifications and description: https://docs.google.com/document/d/1AkH5-G_c3fRktcDeDpMyH0ywgdX_3kZ6Lof_m2RvSQ4/edit

### CODE REVIEW

File: process/factory/shard/interceptorsContainerFactory.go [hunk 1]
```diff
@@ -431,6 +431,16 @@ func (icf *interceptorsContainerFactory) generateMiniBlocksInterceptors() ([]str
...
+	identifierMiniBlocks := factory.MiniBlocksTopic + shardC.CommunicationIdentifier(sharding.MetachainShardId)
+
+	interceptor, err := icf.createOneMiniBlocksInterceptor(identifierMiniBlocks)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	keys[noOfShards] = identifierMiniBlocks
+	interceptorSlice[noOfShards] = interceptor
+
```
Comment (BeniaminDrasovean) [inline process/factory/shard/interceptorsContainerFactory.go]: duplicated code from lines 423-431

File: process/transaction/baseProcess.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+	if acntSnd.GetNonce() < tx.Nonce {
+		return process.ErrHigherNonceInTransaction
+	}
+	if acntSnd.GetNonce() > tx.Nonce {
+		return process.ErrLowerNonceInTransaction
+	}
+
+	cost := big.NewInt(0)
+	cost = cost.Mul(big.NewInt(0).SetUint64(tx.GasPrice), big.NewInt(0).SetUint64(tx.GasLimit))
```
Comment (BeniaminDrasovean) [inline process/transaction/baseProcess.go]: You could use `cost.Mul(big.NewInt(0).SetUint64(tx.GasPrice), big.NewInt(0).SetUint64(tx.GasLimit))` without `cost = `. Also for the next line.

File: sharding/multiShardCoordinator.go [hunk 3]
```diff
@@ -44,6 +44,17 @@ func (msc *multiShardCoordinator) calculateMasks() (uint32, uint32) {
 	return (1 << uint(n)) - 1, (1 << uint(n-1)) - 1
 }
 
+const metaChainIdentifier uint8 = 255
```
Comment (BeniaminDrasovean) [inline sharding/multiShardCoordinator.go]: Move const to the top of the file.

File: process/transaction/baseProcess.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	var acntSrc, acntDst *state.Account
+
+	shardForCurrentNode := txProc.shardCoordinator.SelfId()
+	shardForSrc := txProc.shardCoordinator.ComputeId(adrSrc)
+	shardForDst := txProc.shardCoordinator.ComputeId(adrDst)
+
+	srcInShard := shardForSrc == shardForCurrentNode
+	dstInShard := shardForDst == shardForCurrentNode
+
+	if srcInShard && adrSrc == nil ||
```
Comment (miiu96) [inline process/transaction/baseProcess.go]: adrSrc check if IsInterfaceNil

File: process/transaction/baseProcess.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+		dstInShard && adrDst == nil {
```
Comment (miiu96) [inline process/transaction/baseProcess.go]: adrDst check if IsInterfaceNil

File: process/transaction/baseProcess.go (hunk 2 above)
Comment (miiu96) [inline process/transaction/baseProcess.go]: you can remove "cost = "

File: process/transaction/baseProcess.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+	if acntSnd.GetNonce() < tx.Nonce {
+		return process.ErrHigherNonceInTransaction
+	}
+	if acntSnd.GetNonce() > tx.Nonce {
+		return process.ErrLowerNonceInTransaction
+	}
+
+	cost := big.NewInt(0)
+	cost = cost.Mul(big.NewInt(0).SetUint64(tx.GasPrice), big.NewInt(0).SetUint64(tx.GasLimit))
+	cost = cost.Add(cost, tx.Value)
```
Comment (miiu96) [inline process/transaction/baseProcess.go]: same here

File: sharding/multiShardCoordinator.go [hunk 7]
```diff
@@ -44,6 +44,17 @@ func (msc *multiShardCoordinator) calculateMasks() (uint32, uint32) {
...
+
+func isMetaChainShardId(identifier []byte) bool {
+	for i := 0; i < len(identifier); i++ {
+		if identifier[i] != metaChainIdentifier {
+			return false
+		}
+	}
```
Comment (miiu96) [inline sharding/multiShardCoordinator.go]: add empty line

File: dataRetriever/dataPool/metaDataPool.go [hunk 8]
```diff
@@ -47,8 +59,8 @@ func (mdp *metaDataPool) MetaChainBlocks() storage.Cacher {
 }
 
 // MiniBlockHashes returns the holder for meta mini block hashes
-func (mdp *metaDataPool) MiniBlockHashes() dataRetriever.ShardedDataCacherNotifier {
-	return mdp.miniBlockHashes
+func (mdp *metaDataPool) MiniBlocks() storage.Cacher {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/dataPool/metaDataPool.go]: done

File: dataRetriever/dataPool/metaDataPool.go [hunk 9]
```diff
@@ -62,6 +74,16 @@ func (mdp *metaDataPool) HeadersNonces() dataRetriever.Uint64SyncMapCacher {
 	return mdp.headersNonces
 }
 
+// Transaction returns the holder for transactions which interact the metachain
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/dataPool/metaDataPool.go]: done

File: dataRetriever/factory/metachain/resolversContainerFactory.go [hunk 10]
```diff
@@ -222,6 +266,166 @@ func (rcf *resolversContainerFactory) createMetaChainHeaderResolver(identifier s
...
+		identifierTx := topic + shardC.CommunicationIdentifier(idx)
+		excludePeersFromTopic := topic + shardC.CommunicationIdentifier(shardC.SelfId())
+
+		resolver, err := rcf.createTxResolver(identifierTx, excludePeersFromTopic, unit, dataPool)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		resolverSlice[idx] = resolver
+		keys[idx] = identifierTx
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/factory/metachain/resolversContainerFactory.go]: would not safe lines of codes at all.

File: dataRetriever/factory/metachain/resolversContainerFactory.go [hunk 11]
```diff
@@ -222,6 +266,166 @@ func (rcf *resolversContainerFactory) createMetaChainHeaderResolver(identifier s
...
+		identifierMiniBlocks := factory.MiniBlocksTopic + shardC.CommunicationIdentifier(idx)
+		excludePeersFromTopic := factory.MiniBlocksTopic + shardC.CommunicationIdentifier(shardC.SelfId())
+
+		resolver, err := rcf.createMiniBlocksResolver(identifierMiniBlocks, excludePeersFromTopic)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		resolverSlice[idx] = resolver
+		keys[idx] = identifierMiniBlocks
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/factory/metachain/resolversContainerFactory.go]: would not safe lines of codes at all. already separated the main creation

File: dataRetriever/factory/metachain/resolversContainerFactory.go [hunk 12]
```diff
@@ -222,6 +266,166 @@ func (rcf *resolversContainerFactory) createMetaChainHeaderResolver(identifier s
...
+		rcf.messenger,
+		topic,
+		peerListCreator,
+		rcf.marshalizer,
+		rcf.intRandomizer,
+		uint32(0),
+	)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/factory/metachain/resolversContainerFactory.go]: done

File: dataRetriever/factory/shard/resolversContainerFactory.go [hunk 13]
```diff
@@ -204,6 +204,17 @@ func (rcf *resolversContainerFactory) generateTxResolvers(
...
+	identifierTx := topic + shardC.CommunicationIdentifier(sharding.MetachainShardId)
+	excludePeersFromTopic := topic + shardC.CommunicationIdentifier(shardC.SelfId())
+
+	resolver, err := rcf.createTxResolver(identifierTx, excludePeersFromTopic, unit, dataPool)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	resolverSlice[noOfShards] = resolver
+	keys[noOfShards] = identifierTx
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/factory/shard/resolversContainerFactory.go]: not really. it is a factory.

File: dataRetriever/factory/shard/resolversContainerFactory.go [hunk 14]
```diff
@@ -338,6 +349,17 @@ func (rcf *resolversContainerFactory) generateMiniBlocksResolvers() ([]string, [
 		keys[idx] = identifierMiniBlocks
 	}
 
+	identifierMiniBlocks := factory.MiniBlocksTopic + shardC.CommunicationIdentifier(sharding.MetachainShardId)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/factory/shard/resolversContainerFactory.go]: discucssed

File: process/factory/metachain/interceptorsContainerFactory.go [hunk 15]
```diff
@@ -196,6 +259,118 @@ func (icf *interceptorsContainerFactory) createOneShardHeaderInterceptor(identif
...
+	for idx := uint32(0); idx < noOfShards; idx++ {
+		identifierMiniBlocks := factory.MiniBlocksTopic + shardC.CommunicationIdentifier(idx)
+
+		interceptor, err := icf.createOneMiniBlocksInterceptor(identifierMiniBlocks)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		keys[int(idx)] = identifierMiniBlocks
+		interceptorSlice[int(idx)] = interceptor
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/factory/metachain/interceptorsContainerFactory.go]: discucssed

File: process/factory/metachain/vmContainerFactory_test.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
+package metachain
+
+import (
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/process/factory"
+	"github.com/ElrondNetwork/elrond-go/process/mock"
+	"github.com/stretchr/testify/assert"
+	"testing"
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/factory/metachain/vmContainerFactory_test.go]: done

File: process/factory/shard/interceptorsContainerFactory.go (hunk 1 above)
Comment (sasurobert) [reply process/factory/shard/interceptorsContainerFactory.go]: discussed

File: process/transaction/baseProcess.go (hunk 2 above)
Comment (sasurobert) [reply process/transaction/baseProcess.go]: done

File: sharding/multiShardCoordinator.go (hunk 3 above)
Comment (sasurobert) [reply sharding/multiShardCoordinator.go]: done

File: cmd/node/factory/structs.go [hunk 17]
```diff
@@ -916,6 +917,15 @@ func createMetaChainDataStoreFromConfig(
 					_ = shardHdrHashNonceUnits[i].DestroyUnit()
 				}
 			}
+			if txUnit != nil {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/factory/structs.go]: we do not really use if err == nil, and I would not use it.

File: dataRetriever/dataPool/metaDataPool.go [hunk 18]
```diff
@@ -47,8 +59,8 @@ func (mdp *metaDataPool) MetaChainBlocks() storage.Cacher {
 }
 
 // MiniBlockHashes returns the holder for meta mini block hashes
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/dataPool/metaDataPool.go]: done

File: dataRetriever/dataPool/metaDataPool.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/dataPool/metaDataPool.go]: done

File: dataRetriever/dataPool/metaDataPool_test.go [hunk 19]
```diff
@@ -14,9 +14,11 @@ import (
 func TestNewMetaDataPool_NilMetaBlockShouldErr(t *testing.T) {
 	tdp, err := dataPool.NewMetaDataPool(
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/dataPool/metaDataPool_test.go]: agreed. done.

File: dataRetriever/factory/metachain/resolversContainerFactory.go [hunk 20]
```diff
@@ -222,6 +266,166 @@ func (rcf *resolversContainerFactory) createMetaChainHeaderResolver(identifier s
...
+
+func (rcf *resolversContainerFactory) generateTxResolvers(
+	topic string,
+	unit dataRetriever.UnitType,
+	dataPool dataRetriever.ShardedDataCacherNotifier,
+) ([]string, []dataRetriever.Resolver, error) {
+
+	shardC := rcf.shardCoordinator
+
+	noOfShards := shardC.NumberOfShards()
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/factory/metachain/resolversContainerFactory.go]: noOfShards is used in most of the project, not nr

File: dataRetriever/resolvers/genericBlockBodyResolver.go [hunk 21]
```diff
@@ -23,7 +23,8 @@ func NewGenericBlockBodyResolver(
 	senderResolver dataRetriever.TopicResolverSender,
 	miniBlockPool storage.Cacher,
 	miniBlockStorage storage.Storer,
-	marshalizer marshal.Marshalizer) (*GenericBlockBodyResolver, error) {
+	marshalizer marshal.Marshalizer,
+) (*GenericBlockBodyResolver, error) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply dataRetriever/resolvers/genericBlockBodyResolver.go]: done

File: integrationTests/multiShard/smartContract/testInitializer.go [hunk 22]
```diff
@@ -766,6 +774,18 @@ func createMetaNetNode(
 	store := createTestMetaStore(shardCoordinator)
 	uint64Converter := uint64ByteSlice.NewBigEndianConverter()
 
+	feeHandler := &mock.FeeHandlerStub{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/multiShard/smartContract/testInitializer.go]: done.

File: process/factory/metachain/vmContainerFactory_test.go [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+)
+
+func TestNewVMContainerFactory_NilAccounts(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/factory/metachain/vmContainerFactory_test.go]: done

File: process/factory/metachain/vmContainerFactory_test.go [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+	vmf, err := NewVMContainerFactory(
+		nil,
+		&mock.AddressConverterMock{},
+	)
+
+	assert.Nil(t, vmf)
+	assert.Equal(t, process.ErrNilAccountsAdapter, err)
+}
+
+func TestNewVMContainerFactory_NilAddressConverter(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/factory/metachain/vmContainerFactory_test.go]: done

File: process/factory/shard/interceptorsContainerFactory.go [hunk 25]
```diff
@@ -416,8 +416,8 @@ func (icf *interceptorsContainerFactory) generateHdrInterceptor() ([]string, []p
 func (icf *interceptorsContainerFactory) generateMiniBlocksInterceptors() ([]string, []process.Interceptor, error) {
 	shardC := icf.shardCoordinator
 	noOfShards := shardC.NumberOfShards()
-	keys := make([]string, noOfShards)
-	interceptorSlice := make([]process.Interceptor, noOfShards)
+	keys := make([]string, noOfShards+1)
+	interceptorSlice := make([]process.Interceptor, noOfShards+1)
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/factory/shard/interceptorsContainerFactory.go]: done

File: process/transaction/baseProcess.go (hunk 4 above)
Comment (sasurobert) [reply process/transaction/baseProcess.go]: good catch.

File: process/transaction/baseProcess.go (hunk 5 above)
Comment (sasurobert) [reply process/transaction/baseProcess.go]: good catch

File: process/transaction/baseProcess.go (hunk 2 above)
Comment (sasurobert) [reply process/transaction/baseProcess.go]: done

File: process/transaction/baseProcess.go (hunk 6 above)
Comment (sasurobert) [reply process/transaction/baseProcess.go]: done

File: dataRetriever/factory/metachain/resolversContainerFactory.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply dataRetriever/factory/metachain/resolversContainerFactory.go]: ok

File: dataRetriever/factory/metachain/resolversContainerFactory.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply dataRetriever/factory/metachain/resolversContainerFactory.go]: ok

File: dataRetriever/factory/shard/resolversContainerFactory.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply dataRetriever/factory/shard/resolversContainerFactory.go]: ok

**Pull request merged**